src/software/ai/hl/stp/play/defense_play.cpp

```diff
@@ -97,23 +116,71 @@ void DefensePlay::getNextTactics(TacticCoroutine::push_type &yield)
 
         // Assign ShadowEnemy tactics until we have every enemy covered. If there any
         // extra friendly robots, have them perform a reasonable default defensive tactic
-        for (unsigned i = 0;
-             i < std::min(stop_tactics.size(), shadow_enemy_tactics.size()); i++)
+        if (enemy_threats.size() > 0)
         {
-            if (i < enemy_threats.size())
+            defense_shadow_enemy_tactic->updateWorldParams(
+                world.field(), world.friendlyTeam(), world.enemyTeam(), world.ball());
+            defense_shadow_enemy_tactic->updateControlParams(enemy_threats.at(1));
+            result.emplace_back(defense_shadow_enemy_tactic);
+        }
+        else
+        {
+            // swarm the enemy nearest the ball
+            if (world.enemyTeam().numRobots() > 0)
```

Comment:
Could we break this case out into a function?

Reply:
Yeah I agree with that. The do-while is 100 lines and contains multiple other loops

Reply:
refactored into new function

Reply:
This is still pretty horrific, but I'll leave it for now since we plan to re-visit and test all these.